sw/device/lib/testing/hmac_testutils.c

```diff
@@ -0,0 +1,84 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/testing/hmac_testutils.h"
+
+#include "sw/device/lib/dif/dif_hmac.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/testing/check.h"
+
+void hmac_testutils_check_message_length(const dif_hmac_t *hmac,
+                                         uint64_t expected_sent_bits) {
+  uint64_t sent_bits;
+  CHECK_DIF_OK(dif_hmac_get_message_length(hmac, &sent_bits));
+
+  // 64bit formatting is not supported, so split into hi and lo hex 32bit
+  // values. These should appear as 64bit hex values in the debug output.
+  CHECK(expected_sent_bits == sent_bits,
+        "Message length mismatch. "
+        "Expected 0x%08x%08x bits but got 0x%08x%08x bits.",
+        (uint32_t)(expected_sent_bits >> 32), (uint32_t)expected_sent_bits,
+        (uint32_t)(sent_bits >> 32), (uint32_t)sent_bits);
+}
+
+/**
+ * Checks whether the HMAC FIFO is empty.
+ */
+static bool check_fifo_empty(const dif_hmac_t *hmac) {
+  uint32_t fifo_depth;
+  CHECK_DIF_OK(dif_hmac_fifo_count_entries(hmac, &fifo_depth));
+  return fifo_depth == 0;
+}
+void hmac_testutils_fifo_empty_polled(const dif_hmac_t *hmac) {
+  IBEX_SPIN_FOR(check_fifo_empty(hmac), HMAC_TESTUTILS_FIFO_EMPTY_TIMEOUT_USEC);
+}
+
+static bool check_finished(const dif_hmac_t *hmac,
+                           dif_hmac_digest_t *digest_out) {
+  dif_result_t res = dif_hmac_finish(hmac, digest_out);
+  CHECK(res == kDifOk || res == kDifUnavailable, "HMAC error = %d", res);
+
+  return res == kDifOk;
+}
+void hmac_testutils_finish_polled(const dif_hmac_t *hmac,
+                                  dif_hmac_digest_t *digest_out) {
+  IBEX_SPIN_FOR(check_finished(hmac, digest_out),
+                HMAC_TESTUTILS_FINISH_TIMEOUT_USEC);
+}
+
+void hmac_testutils_check_digest(const dif_hmac_digest_t *expected,
+                                 const dif_hmac_digest_t *actual) {
+  for (int i = 0; i < 8; ++i) {
```

Comment:
I would use `ARRAYSIZE(actual->digest)` here.

Reply:
Done in #10995